### PR TITLE
Add mana semantic validation

### LINKED ISSUE #1692: Add mana-related semantic validations
- [ ] https://github.com/iotaledger/iota-sdk/pull/1762/files#diff-0cd9fdcd945509b7f15ba76e36be88c3db111d6c0967dd15327efe148d5a9421R313

### PR DESCRIPTION
**Pull request by DaughterOfMars**
# Description of change

Validates transaction mana usage semantics.

## Links to any relevant issues

Closes #1692

### CODE REVIEW

File: sdk/src/types/block/semantic.rs [hunk 1]
```diff
@@ -470,9 +513,16 @@ impl<'a> SemanticValidationContext<'a> {
...
+        if self.input_mana != self.output_mana {
+            if self.input_mana > self.output_mana
+                && !self.transaction.has_capability(TransactionCapabilityFlag::BurnMana)
+            {
+                return Ok(Some(
+                    TransactionFailureReason::TransactionCapabilityManaBurningNotAllowed,
+                ));
+            } else {
+                return Ok(Some(TransactionFailureReason::InvalidManaAmount));
+            }
```
Comment (Alex6323) [inline sdk/src/types/block/semantic.rs]: hmm, is that correct? what in case the transaction has the capability? Still an error?

Comment (DaughterOfMars) [reply sdk/src/types/block/semantic.rs]: pretty sure it's right

Comment (Alex6323) [reply sdk/src/types/block/semantic.rs]: hmm, according to [TIP-39](https://github.com/iotaledger/tips/blob/tip39/tips/TIP-0039/tip-0039.md#mana-transaction-validation-rules) the transaction should be valid if the transaction can burn mana.  @thibault-martinez , right?

```rust
if self.input_mana != self.output_mana {
    if self.input_mana > self.output_mana {
          if !self.transaction.has_capability(TransactionCapabilityFlag::BurnMana) {
               return Ok(Some(TransactionFailureReason::TransactionCapabilityManaBurningNotAllowed));
          } 
    } else {
          return Ok(Some(TransactionFailureReason::InvalidManaAmount));
    }
}
```

Comment (thibault-martinez) [reply sdk/src/types/block/semantic.rs]: I don't review draft PRs :trollface:

Comment (thibault-martinez) [reply sdk/src/types/block/semantic.rs]: I have slightly changed it

File: sdk/src/types/block/semantic/mod.rs [hunk 2]
```diff
@@ -160,7 +162,36 @@ impl<'a> SemanticValidationContext<'a> {
                 .checked_add(amount)
                 .ok_or(Error::ConsumedAmountOverflow)?;
 
-            self.input_mana = self.input_mana.checked_add(mana).ok_or(Error::ConsumedManaOverflow)?;
+            let potential_mana = {
+                let min_deposit = consumed_output.minimum_amount(self.protocol_parameters.storage_score_parameters());
```
Comment (Thoralf-M) [inline sdk/src/types/block/semantic/mod.rs:167]:
```suggestion
                // Deposit amount doesn't generate mana
                let min_deposit = consumed_output.minimum_amount(self.protocol_parameters.storage_score_parameters());
```

**Review by Thoralf-M — DISMISSED** [review]

**Review by Thoralf-M — APPROVED** [review]

**Review by kwek20 — APPROVED** [review]

**Pull request merged**